packages/xrpl/test/integration/transactions/batch.test.ts

```diff
@@ -0,0 +1,113 @@
+import serverUrl from '../serverUrl'
+import {
+  setupClient,
+  teardownClient,
+  type XrplIntegrationTestContext,
+} from '../setup'
+import {
+  generateFundedWallet,
+  testTransaction,
+  verifySubmittedTransaction,
+} from '../utils'
+
+// how long before each test case times out
+const TIMEOUT = 20000
+
+describe('Batch', function () {
+  let testContext: XrplIntegrationTestContext
+  let destination: Wallet
+  let wallet2: Wallet
+
+  async function testBatchTransaction(
+    batch: Batch,
+    wallet: Wallet,
+    retry?: {
+      count: number
+      delayMs: number
+    },
+  ): Promise<void> {
+    await testTransaction(testContext.client, batch, wallet, retry)
+    const promises: Array<Promise<void>> = []
+    for (const rawTx of batch.RawTransactions) {
+      promises.push(
+        verifySubmittedTransaction(testContext.client, rawTx.RawTransaction),
+      )
+    }
+    await Promise.all(promises)
+  }
+
+  beforeAll(async () => {
+    testContext = await setupClient(serverUrl)
+    wallet2 = await generateFundedWallet(testContext.client)
+    destination = await generateFundedWallet(testContext.client)
+  }, TIMEOUT)
+  afterAll(async () => teardownClient(testContext))
+
+  it(
+    'base',
+    async () => {
+      const tx: Batch = {
+        TransactionType: 'Batch',
+        Account: testContext.wallet.classicAddress,
+        Flags: BatchFlags.tfAllOrNothing,
+        RawTransactions: [
+          {
+            RawTransaction: {
+              TransactionType: 'Payment',
+              Account: testContext.wallet.classicAddress,
+              Destination: destination.classicAddress,
+              Amount: '10000000',
+            },
+          },
+          {
+            RawTransaction: {
+              TransactionType: 'Payment',
+              Account: testContext.wallet.classicAddress,
+              Destination: destination.classicAddress,
+              Amount: '10000000',
+            },
+          },
+        ],
+      }
+      const autofilled = await testContext.client.autofill(tx)
+      await testBatchTransaction(autofilled, testContext.wallet)
+    },
+    TIMEOUT,
+  )
+
+  it(
+    'batch multisign',
+    async () => {
+      const tx: Batch = {
+        TransactionType: 'Batch',
+        Account: testContext.wallet.classicAddress,
+        Flags: BatchFlags.tfAllOrNothing,
+        RawTransactions: [
+          {
+            RawTransaction: {
+              TransactionType: 'Payment',
+              Account: testContext.wallet.classicAddress,
+              Destination: destination.classicAddress,
+              Amount: '10000000',
+            },
+          },
+          {
+            RawTransaction: {
+              TransactionType: 'Payment',
+              Account: wallet2.classicAddress,
+              Destination: destination.classicAddress,
+              Amount: '10000000',
+            },
+          },
+        ],
+      }
+      const autofilled = await testContext.client.autofill(tx, 2)
```

Comment:
```suggestion
      const autofilled = await testContext.client.autofill(tx)
```

Do Batch transactions need to manually specify the number of signers? `calculateFeePerTransactionType` does not consider `signersCount` whilst computing the transactions fees for `Batch` transaction.

Reply:
I'm using `signersCount` as the total number of multisign signers _and_ the total number of Batch signers.

Reply:
I don't understand three aspects of the fee calculation:
1. The [specification (Section 2.1)](https://github.com/XRPLF/XRPL-Standards/discussions/162) states that the fees of the inner-Batch transactions are incorporated. How does the implementation ensure that? Suppose the inner batch transaction contains AMMCreate, AccountDelete (or) EscrowFinish transactions, should't their fees be computed in a recursive/iterative manner by going through all the inner transactions?

2. Based on my understanding of the spec, the appropriate fees would be:
```
// The definition of baseFee follows from the existing code.
const n = (tx.BatchSigners?.length ?? 0)
fee_accumulator = (n + 2) * baseFee

for await (inner_txn in inner_transactions) {
    fee_accumulator += calculateFeePerTransactionType(client, inner_txn, <I don't know the signerCount value>)
}
```

My understanding is in-line with [this comment of yours](https://github.com/XRPLF/XRPL-Standards/discussions/162#discussioncomment-7800870). Has the spec changed?

3. If there are multiple signers, then the trasaction pays for it in two factors: the length of BatchSigners array and the signersCount. Is that accurate? Is this explained in the spec?

The UX is not good. The `signersCount` could have been auto-filled by the library. There is no need for users to specify that.

<img width="1000" alt="image" src="https://github.com/user-attachments/assets/11169909-5397-482d-9905-9d66074955eb" />
<img width="654" alt="image" src="https://github.com/user-attachments/assets/d2efffe8-5d92-47e2-aaef-91a08d9bec4b" />

Reply:
> 1. The [specification (Section 2.1)](https://github.com/XRPLF/XRPL-Standards/discussions/162) states that the fees of the inner-Batch transactions are incorporated. How does the implementation ensure that? Suppose the inner batch transaction contains AMMCreate, AccountDelete (or) EscrowFinish transactions, should't their fees be computed in a recursive/iterative manner by going through all the inner transactions?

Good catch - fixed.

> 3. If there are multiple signers, then the trasaction pays for it in two factors: the length of BatchSigners array and the signersCount. Is that accurate? Is this explained in the spec?
> 
> The UX is not good. The `signersCount` could have been auto-filled by the library. There is no need for users to specify that.

`signersCount` cannot be autofilled, because if those are filled in, the transaction has already been signed. `autofill` has to run before the transaction is signed.

Reply:
> signersCount cannot be autofilled, because if those are filled in, the transaction has already been signed. autofill has to run before the transaction is signed.

Okay, that makes sense.